File: Application/src/main/java/by/gastrofest/service/GastrofestService.java
package by.gastrofest.service;

import by.gastrofest.dbo.one.GastroFestDbo;
import by.gastrofest.repository.one.GastrofestRepository;
import lombok.RequiredArgsConstructor;
import org.jsoup.nodes.Document;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;

import static by.gastrofest.constant.GastrofestConstants.LOCATION_CLASS;
import static by.gastrofest.constant.GastrofestConstants.MAIN_POST_CLASS;
import static by.gastrofest.constant.GastrofestConstants.TITLE_CLASS;
import static by.gastrofest.constant.MainConstants.IMG_TAG;
import static by.gastrofest.constant.MainConstants.SRC_PROPERTY;
import static by.gastrofest.utils.HttpUtil.getEncodedString;

@Service
@RequiredArgsConstructor

public class GastrofestService {

    private final GastrofestRepository repository;

    @Transactional("transactionManager")
    public GastroFestDbo save(final GastroFestDbo gastrofestDbo) {
        return repository.findByStartDate(gastrofestDbo.getStartDate())
                .orElseGet(() -> repository.save(gastrofestDbo));
    }

    public GastroFestDbo extractGastrofestFromElement(final Document document) {
        final var mainElement = document.getElementsByClass(MAIN_POST_CLASS).get(0);
        final var gastrofestName = mainElement.getElementsByClass(TITLE_CLASS).get(0).text();
        final var imageLink = mainElement.getElementsByTag(IMG_TAG).get(0).absUrl(SRC_PROPERTY);
        final var imageBase64 = getEncodedString(imageLink);
        final var locationText = mainElement.getElementsByClass(LOCATION_CLASS).get(0).text();
        final var locations = locationText.split("\\|")[0].trim();
        final var dates = locationText.split("\\|")[1].trim().split(" - ");
        final var startDate = parseDate(dates, 0);
        final var endDate = parseDate(dates, 1);
        return new GastroFestDbo(gastrofestName, locations, imageLink, imageBase64, startDate, endDate);
    }

    private LocalDate parseDate(final String[] dates, final int range) {
        return LocalDate.of(
                LocalDate.now().getYear(),
                Integer.parseInt(dates[range].split("\\.")[1]),
                Integer.parseInt(dates[range].split("\\.")[0]));
    }
}

File: Parser/src/main/kotlin/by/gastrofest/parser/service/GastrofestService.kt
package by.gastrofest.parser.service

import by.gastrofest.parser.constant.GASTROFEST_TITLE_CLASS
import by.gastrofest.parser.constant.LOCATION_CLASS
import by.gastrofest.parser.constant.MAIN_POST_CLASS
import by.gastrofest.parser.constant.SOURCE_TAG
import by.gastrofest.parser.constant.SRC_PROPERTY
import by.gastrofest.parser.constant.VIDEO_TAG
import by.gastrofest.parser.model.GastroFest
import org.jsoup.nodes.Document
import java.time.LocalDate
import java.time.Year
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatterBuilder
import java.time.temporal.ChronoField
import java.util.Locale

private val RUSSIAN_DATE_FORMATTER: DateTimeFormatter = DateTimeFormatterBuilder()
    .parseCaseInsensitive()
    .appendPattern("d MMMM")
    .parseDefaulting(ChronoField.YEAR, Year.now().value.toLong())
    .toFormatter(Locale("ru"))

fun extractGastrofestFromElement(document: Document): GastroFest {
    val mainElement = document.getElementsByClass(MAIN_POST_CLASS)[0]
    val gastrofestName = mainElement.getElementsByClass(GASTROFEST_TITLE_CLASS)[0].text()
    val imageLink = mainElement.getElementsByTag(VIDEO_TAG)[0].getElementsByTag(SOURCE_TAG)[0].absUrl(SRC_PROPERTY)
    val locationTexts = mainElement.getElementsByClass(LOCATION_CLASS)[0].text().split("\\|".toRegex())
    val locations = locationTexts[0].trim()
    val dates = locationTexts[1].trim().split(" - ")
    val startDate = parseDate(dates, 0)
    val endDate = parseDate(dates, 1)
    return GastroFest(null, gastrofestName, imageLink, locations, startDate, endDate)
}

private fun parseDate(dates: List<String>, range: Int): LocalDate =
    LocalDate.parse(dates[range].trim(), RUSSIAN_DATE_FORMATTER)
